src/common/ad_mem.c

```diff
@@ -3,5 +3,1578 @@
+			tmp = sorter[++i];
+			if (tmp->gp_unit != gd->gd_unit)
+				break;
+
+			D_ASSERT(tmp->gd_unit >= gd->gd_unit);
+			if (group_unit_avail(tmp) > avail ||
+			    tmp->gd_addr > gd->gd_addr)
+				break;
+
+			sorter[pos] = tmp;
+			sorter[i] = grp;
+			pos = i;
+		}
+	}
+}
+
+/** add a group to arena */
+static void
+arena_add_grp(struct ad_arena *arena, struct ad_group *grp, int *pos)
+{
+	struct ad_group_df **grps_size = arena->ar_size_sorter;
+	struct ad_group_df **grps_addr = arena->ar_addr_sorter;
+	struct ad_group_df  *gd   = grp->gp_df;d
+	struct ad_group_df  *tmp  = NULL;
+	int		     avail;
+	int		     start;
+	int		     end;
+	int		     cur;
+	int		     len;
+	int		     i;
+
+	/* no WAL, in DRAM */
+	len = arena->ar_grp_nr++;
+	if (len == 0) {
+		grps_addr[0] = grp;
+		grps_size[0] = grp;
+		if (pos)
+			*pos = 0;
+		return;
+	}
+
+	/* step-1: add the group the address sorter */
+	for (start = 0, end = len - 1; start <= end) {
+		cur = (start + end) / 2;
+		tmp = grps_addr[cur];
+		if (tmp->gd_addr < grp->gd_addr)
+			rc = -1;
+		else if (tmp->gd_addr < grp->gd_addr)
+			rc = 1;
+		else
+			D_ASSERT(0);
+
+		if (rc < 0)
+			start = cur + 1;
+		else
+			end = cur - 1;
+        }
+	if (tmp->gd_addr < gd->gd_addr)
+		cur += 1;
+	if (cur < len)
+		memmove(grps_addr[cur], grps_addr[cur + 1], (len - cur) * sizeof(grps_addr[0]));
+
+	/* step-2: add the group the size sorter */
+	avail = group_unit_avail(gd);
+	for (start = 0, end = len - 1; start <= end) {
+		cur = (start + end) / 2;
+		tmp = grps_size[cur];
+		if (tmp->gd_unit == grp->gd_unit) {
+			int	n = group_unit_avail(tmp);
+
+			if (n < avail) {
+				rc = -1;
+			else (n > avail)
+				rc = 1;
+			else
+				rc = (tmp->gd_addr < gd->gd_addr) ? -1 : 1;
+
+		} else if (tmp->gd_unit < grp->gd_unit) {
+			rc = -1;
+		} else {
+			rc = 1;
+		}
+
+		if (rc < 0)
+			start = cur + 1;
+		else
+			end = cur - 1;
+        }
+
+	if (tmp->gd_unit < grp->gd_unit ||
+	    (tmp->gd_unit == grp->gd_unit && tmp->gd_addr < gd->gd_addr))
+		cur += 1;
+
+	if (cur < len)
+		memmove(grps_size[cur], grps_size[cur + 1], (len - cur) * sizeof(grps_size[0]));
+
+	if (pos)
+		*pos = cur;
+	grps_size[cur] = grp;
+}
+
+static int
+find_bits(uint64_t *used, uint64_t *reserved, int bmap_sz, int bits_min, int *bits)
+{
+	int	nr_saved;
+	int	at_saved;
+	int	nr;
+	int	at;
+	int	i;
+	int	j;
+
+	nr = nr_saved = 0;
+	at = at_saved = -1;
+
+	for (i = n = 0; i < bmap_sz; i++) {
+		uint64_t free_bits = ~used[i];
+		uint64_t tmp;
+
+		if (reserved)
+			free_bits |= ~reserved[i];
+
+		if (free_bits == 0) { /* no space at all */
+			if (nr > nr_saved) {
+				nr_saved = nr;
+				at_saved = at;
+			}
+			nr = 0;
+			at = -1;
+			continue;
+		}
+
+		j = ffsll(free_bits);
+		D_ASSERT(j > 0);
+		if (at >= 0 && j == 1) {
+			D_ASSERT(nr > 0);
+			nr++;
+		} else {
+			at = i * 64 + j - 1;
+			nr = 1;
+		}
+
+		for (; j < 64; j++) {
+			if (nr == *bits) /* done */
+				goto out;
+
+			if (isset(free_bits, j)) {
```

Comment:
isset/isclr/setbit/clrbit maros seems only works with "uint8 []" type (see /usr/include/sys/param.h),
if use this macros, may define all bitmap to be "uint8_t []" or "char []".

Reply:
correct, we need to do some typecast at here